### ISSUE TITLE
LAMMPS `develop` triggers assertion when `kokkosp_allocate_data` called with nullptr and size 0

### ISSUE THREAD

**Issue by cwpearson** [kokkos/kokkos-tools#256]
https://github.com/kokkos/kokkos-tools/blob/58258bd6666dc6d11cb748441f62e68f9d22ba27/profiling/space-time-stack/kp_space_time_stack.cpp#L525

I don't know anything about LAMMPS, but I observe it triggering `kokkosp_allocate_data` with nullptr and size 0 allocation twice, which triggers this assertion. Interest in a PR that prints a warning in that specific case, but doesn't abort?

`patch_17Apr2024` release in LAMMPS also has this problem. `stable_2Aug2023_update3` does not trigger this problem.

@simongdg

**Comment by masterleinad** [comment]
> I observe it triggering kokkosp_allocate_data with nullptr and size 0 allocation twice, which triggers this assertion. 

Do you know what `Kokkos` configuration triggers it?

**Comment by stanmoore1** [comment]
I just ran into this as well:
```
lmp_vortex_kokkos: kokkos-tools/profiling/space-time-stack/kp_space_time_stack.cpp:525: void KokkosTools::SpaceTimeStack::Allocations::allocate(std::string&&, const void*, uint64_t, KokkosTools::SpaceTimeStack::StackNode*): Assertion `res.second' failed.
```

Running with CUDA on V100.

**Comment by stanmoore1** [comment]
Here is a full stack trace:
```
(gdb) where
#0  0x00007fffe7b7fcb0 in raise () from /lib64/libc.so.6
#1  0x00007fffe7b8200c in abort () from /lib64/libc.so.6
#2  0x00007fffe7b757d4 in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007fffe7b758c4 in __assert_fail () from /lib64/libc.so.6
#4  0x00007fffdc0baa58 in allocate (frame=<optimized out>, size=0, ptr=0x0,
    name=<unknown type in kokkos-tools/profiling/space-time-stack/kp_space_time_stack.so, CU 0x0, DIE 0xa95c5>, this=0x18de5c50)
    at kokkos-tools/profiling/space-time-stack/kp_space_time_stack.cpp:776
#5  allocate (size=0, ptr=0x0,
    name=0x2425a5d8 "Kokkos::CudaSpace::TeamScratchMemory",
    space=<optimized out>, this=0x18de5b90)
    at kokkos-tools/profiling/space-time-stack/kp_space_time_stack.cpp:772
#6  KokkosTools::SpaceTimeStack::kokkosp_allocate_data (handle=...,
    name=0x2425a5d8 "Kokkos::CudaSpace::TeamScratchMemory", ptr=0x0, size=0)
    at kokkos-tools/profiling/space-time-stack/kp_space_time_stack.cpp:854
#7  0x00007fffdc0bab00 in kokkosp_allocate_data (space=...,
    label=0x2425a5d8 "Kokkos::CudaSpace::TeamScratchMemory", ptr=0x0, size=0)
    at kokkos-tools/profiling/space-time-stack/kp_space_time_stack.cpp:941
#8  0x000000001191c8e4 in invoke_kokkosp_callback<void (*)(Kokkos_Profiling_SpaceHandle, char const*, void const*, unsigned long), Kokkos_Profiling_SpaceHandle const&, char const*, void const*&, unsigned long const&> (
    may_require_global_fencing=Kokkos::Tools::Experimental::No,
    callback=@0x184d7fd0: 0x7fffdc0baab0 <kokkosp_allocate_data(Kokkos::Tools::SpaceHandle, char const*, void const*, uint64_t)>)
    at ../../lib/kokkos/core/src/impl/Kokkos_Profiling.cpp:411
#9  invoke_kokkosp_callback<void (*)(Kokkos_Profiling_SpaceHandle, char const*, void const*, unsigned long), Kokkos_Profiling_SpaceHandle const&, char const*, void const*&, unsigned long const&> (
    callback=@0x184d7fd0: 0x7fffdc0baab0 <kokkosp_allocate_data(Kokkos::Tools::SpaceHandle, char const*, void const*, uint64_t)>,
    may_require_global_fencing=Kokkos::Tools::Experimental::No)
    at ../../lib/kokkos/core/src/impl/Kokkos_Profiling.cpp:280
#10 allocateData (size=0, ptr=0x0, label=..., space=...)
    at ../../lib/kokkos/core/src/impl/Kokkos_Profiling.cpp:411
#11 Kokkos::Profiling::allocateData (handle=...,
    name="Kokkos::CudaSpace::TeamScratchMemory", data=0x0, size=0)
    at ../../lib/kokkos/core/src/impl/Kokkos_Profiling.cpp:1018
#12 0x0000000011956f1c in impl_allocate_common (stream=<optimized out>,
    stream_sync_only=false, arg_handle=..., arg_logical_size=<optimized out>,
    arg_alloc_size=0,
    arg_label=0x11a9e8a0 "Kokkos::CudaSpace::TeamScratchMemory",
    device_id=<optimized out>)
    at ../../lib/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:216
#13 impl_allocate (arg_handle=..., arg_logical_size=<optimized out>,
    arg_alloc_size=0,
    arg_label=0x11a9e8a0 "Kokkos::CudaSpace::TeamScratchMemory",
    this=<optimized out>)
    at ../../lib/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:226
#14 Kokkos::CudaSpace::allocate (this=<optimized out>,
    arg_label=0x11a9e8a0 "Kokkos::CudaSpace::TeamScratchMemory",
    arg_alloc_size=0, arg_logical_size=<optimized out>)
    at ../../lib/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:171
#15 0x000000001194bbdc in Kokkos::Impl::CudaInternal::resize_team_scratch_space
    (this=<optimized out>, scratch_pool_id=<optimized out>, bytes=0,
    force_shrink=<optimized out>)
    at ../../lib/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:434
#16 0x00000000111e6ee0 in Kokkos::Impl::ParallelFor<BuildBorderListFunctor<Kokkos::Cuda>, Kokkos::TeamPolicy<Kokkos::Cuda>, Kokkos::Cuda>::ParallelFor (
    this=0x7fffffff9250, arg_functor=..., arg_policy=...)
    at ../../lib/kokkos/core/src/impl/Kokkos_SharedAlloc.hpp:590
#17 0x00000000111ec71c in parallel_for<Kokkos::TeamPolicy<Kokkos::Cuda>, BuildBorderListFunctor<Kokkos::Cuda> > (functor=..., policy=..., str="")
    at ../../lib/kokkos/core/src/impl/Kokkos_SharedAlloc.hpp:139
#18 parallel_for<Kokkos::TeamPolicy<Kokkos::Cuda>, BuildBorderListFunctor<Kokkos::Cuda> > (functor=..., policy=...)
    at ../../lib/kokkos/core/src/Kokkos_Parallel.hpp:153
#19 LAMMPS_NS::CommKokkos::borders_device<Kokkos::Cuda> (this=0x18de2af0)
    at ../comm_kokkos.cpp:1211
#20 0x00000000111dac38 in LAMMPS_NS::CommKokkos::borders (this=0x18de2af0)
    at ../comm_kokkos.cpp:1071
#21 0x0000000010a18790 in LAMMPS_NS::VerletKokkos::setup (this=0x23e866c0,
    flag=<optimized out>) at ../verlet_kokkos.cpp:105
#22 0x000000001014ef0c in LAMMPS_NS::Run::command (this=0x23e76760,
    narg=<optimized out>, arg=0x23e76a10) at ../run.cpp:171
#23 0x00000000100ad640 in LAMMPS_NS::Input::execute_command (this=0x18de5f40)
    at ../input.cpp:868
#24 0x00000000100ae4d0 in LAMMPS_NS::Input::file (this=0x18de5f40)
    at ../input.cpp:313
#25 0x00000000100211e0 in main (argc=<optimized out>, argv=<optimized out>)
    at ../main.cpp:77
```

**Comment by vlkale** [comment]
Thanks @cwpearson and @stanmoore1  

I don't think I see what Kokkos configuration version you used (as @masterleinad was asking)? 

Looking at git blame:  @ibaned, thanks for the contribution. Do you remember/know why an assert was put here? (I know this was a very long time back). 

A broad comment/opinion on assertions in a Kokkos Tool library: I think a Kokkos Tool library for profiling and debugging (Space-time-stack, simple-kernel-timer, kernel-logger) should only provide advisory/diagnostic information, e.g., a warning printed to screen, about their Kokkos program's execution (which includes the Kokkos Tool library execution itself). 
In other words, a profiling tool should not abort a Kokkos program. With that said, I currently don't see a good alternative to the assert for MPI ranks at this line: 

https://github.com/kokkos/kokkos-tools/blob/58258bd6666dc6d11cb748441f62e68f9d22ba27/profiling/space-time-stack/kp_space_time_stack.cpp#L562


@cwpearson, @ all: If opening a PR for a warning in that specific case, the warning should be printed with `std::cout` rather than printf - this will allow for easier testing with Google Test.

**Comment by crtrott** [comment]
The tool needs to be fixed to handle nulltptr / size==0. This is likely happening for internal scratch buffers or so in Kokkos where we call raw tools callbacks. And this might trigger nulltpr/0 for scratch pads which are zero size - at least that is my suspicion.

I.e. the tool should have something like this:

```c++
if(ptr == nullptr) { assert(size==0); return; }
``` 

or something to that effect.
